Cabal/Distribution/Simple/GHC/Internal.hs

```diff
@@ -432,19 +443,100 @@ profDetailLevelFlag forLib mpl =
 --
-showArchString :: Arch -> String
-showArchString PPC   = "powerpc"
-showArchString PPC64 = "powerpc64"
-showArchString other = display other
+ghcArchString :: Arch -> String
+ghcArchString PPC   = "powerpc"
+ghcArchString PPC64 = "powerpc64"
+ghcArchString other = display other
 
 -- | GHC's rendering of it's host or target 'OS' as used in its platform
 -- strings and certain file locations (such as user package db location).
 --
-showOsString :: OS -> String
-showOsString Windows = "mingw32"
-showOsString OSX     = "darwin"
-showOsString Solaris = "solaris2"
-showOsString other   = display other
+ghcOsString :: OS -> String
+ghcOsString Windows = "mingw32"
+ghcOsString OSX     = "darwin"
+ghcOsString Solaris = "solaris2"
+ghcOsString other   = display other
+
+-- | GHC's rendering of it's platform and compiler version string as used in
+-- certain file locations (such as user package db location).
+-- For example @x86_64-linux-7.10.4@
+--
+ghcPlatformAndVersionString :: Platform -> Version -> String
+ghcPlatformAndVersionString (Platform arch os) version =
+    intercalate "-" [ ghcArchString arch, ghcOsString os, display version ]
+
+
+-- -----------------------------------------------------------------------------
+-- Constructing GHC environment files
+
+-- | The kinds of entries we can stick in a @.ghc.environment@ file.
+--
+data GhcEnvironmentFileEntry =
+       GhcEnvFileComment   String     -- ^ @-- a comment@
+     | GhcEnvFilePackageId UnitId     -- ^ @package-id foo-1.0-4fe301a...@
+     | GhcEnvFilePackageDb PackageDB  -- ^ @global-package-db@,
+                                      --   @user-package-db@ or
+                                      --   @package-db blah/package.conf.d/@
+     | GhcEnvFileClearPackageDbStack  -- ^ @clear-package-db@
+
+-- | Make entries for a GHC environment file based on a 'PackageDBStack' and
+-- a bunch of package (unit) ids.
+--
+-- If you need to do anything more complicated then either use this as a basis
+-- and add more entries, or just make all the entries directly.
+--
+simpleGhcEnvironmentFile :: PackageDBStack
+                         -> [UnitId]
+                         -> [GhcEnvironmentFileEntry]
+simpleGhcEnvironmentFile packageDBs pkgids =
+    GhcEnvFileClearPackageDbStack
+  : map GhcEnvFilePackageDb packageDBs
+ ++ map GhcEnvFilePackageId pkgids
+
+-- | Write a @.ghc.environment-$arch-$os-$ver@ file in the given directory.
+--
+-- The 'Platform' and GHC 'Version' are needed as part of the file name.
+--
+writeGhcEnvironmentFile :: FilePath  -- ^ directory in which to put it
+                        -> Platform  -- ^ the GHC target platform
+                        -> Version   -- ^ the GHC version
+                        -> [GhcEnvironmentFileEntry] -- ^ the content
+                        -> IO ()
+writeGhcEnvironmentFile directory platform ghcversion =
+    writeFileAtomic envfile . BS.pack . renderGhcEnvironmentFile
```

Comment:
It is regretful but understandable why you didn't construct a ByteString in the first place ;)